### ISSUE TITLE
Bug in calling model.eval() in PPO

### ISSUE THREAD

**Issue by idanshen** [huggingface/trl#1569]
When performing PPO step, the code perform the forward pass in [line 798](https://github.com/huggingface/trl/blob/main/trl/trainer/ppo_trainer.py) using the function "batched_forward_pass".
However, "batched_forward_pass" put the model in eval mode ([line 986](https://github.com/huggingface/trl/blob/6d1cb85e73e990c5befac09988528996e8261080/trl/trainer/ppo_trainer.py#L986C1-L986C21)):

        model.eval()

I'm pretty sure this is not intended and can lead to bugs.

**Comment by B-Gendron** [comment]
I was suspicious about this at first sight but finally I think this is normal. Here is how I would explain that:

If you look at `step()` method in `PPOTrainer` class, you see there are two main parts in this training. The first one performs forward path on language model and reference language model, hence there is no backprop made at this stage since we just want the outputs to apply the policy. In a second phase, starting line 807, `train_minibatch()` method is called and here the model is in train mode, which is logical because the loss is computed here. 

Hope this helps!
